### PR TITLE
Added Authentication

### LINKED ISSUE #37: Authorization: Protect remaining queries/mutations
## Parent
- https://github.com/IT-REX-Platform/gits_backend/issues/12

## Acceptance Critera
- [x] Media Service: Queries & mutations are protected by checking if user has necessary permissions (either global or course-specific) to perform action
- [x] Course Service: Added global permissions check for queries/mutations that need it (e.g. createCourse)
- [x] Reward Service: Protect queries & mutations or make internal only
- [x] Skilllevel Service: Protect queries & mutations or make internal only

### PR DESCRIPTION
**Pull request by ZombieAlienRobot**
## Description of changes
Added Authentication for all Controllers.

  
## How has this been tested?

Please describe the test strategy you followed. If not tested, please explain why.

- [X] automated unit test
- [X] automated integration test
- [X] manual, exploratory test

In case of manual test, please document the test well including a set of user instructions and prerequisites. Each including an action, it's result, and where appropriate a screenshot.
    
## Checklist before requesting a review

- [ ] My code is easy to understand
- [ ] My code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [ ] My code fulfills all acceptance criteria
- [ ] The test coverage (line and branch) is reasonably high, especially on Service classes and other classes with complex logic
- [ ] I have made corresponding changes to the documentation or
  the [wiki/adr](https://github.com/IT-REX-Platform/wiki/tree/main/adr)
- [ ] I made no breaking changes in the database schema or if so, I will perform a database migration

## Checklist for reviewer

- The code is easy to understand
- The code follows
  the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of
  this project
- The code is tested or if not, the reason is documented or discussed
- The added and existing tests reasonably cover the code change
- The code has no breaking changes in the database schema or if so, the assignee is aware of it and
  will perform a database migration

### CODE REVIEW

File: src/main/java/de/unistuttgart/iste/gits/media_service/controller/MediaController.java [hunk 1]
```diff
@@ -148,14 +180,86 @@ private boolean uploadUrlInSelectionSet(final DataFetchingEnvironment env) {
...
+        NoAccessToCourseException noAccessToCourseException = null;
+        if (mediaRecord.getCourseIds().isEmpty()) {
+            return;
+        }
+        for (final UUID courseId : mediaRecord.getCourseIds()) {
+            try {
+                validateUserHasAccessToCourse(currentUser, role, courseId);
+                break;
+            } catch (final NoAccessToCourseException exception) {
+                noAccessToCourseException = exception;
```
Comment (PaulBredl) [inline src/main/java/de/unistuttgart/iste/gits/media_service/controller/MediaController.java:237]: Just throw the exception here directly

File: src/main/java/de/unistuttgart/iste/gits/media_service/controller/MediaController.java [hunk 2]
```diff
@@ -90,6 +94,11 @@ public MediaRecord createMediaRecord(@Argument final List<UUID> courseIds,
                                          @Argument final CreateMediaRecordInput input,
                                          @ContextValue final LoggedInUser currentUser,
                                          final DataFetchingEnvironment env) {
+        validateUserHasGlobalPermission(currentUser, Set.of(LoggedInUser.RealmRole.COURSE_CREATOR));
```
Comment (PaulBredl) [inline src/main/java/de/unistuttgart/iste/gits/media_service/controller/MediaController.java:97]: only check if list of course ids is empty

File: src/main/java/de/unistuttgart/iste/gits/media_service/controller/MediaController.java [hunk 3]
```diff
@@ -25,63 +37,55 @@ public class MediaController {
...
-    public List<MediaRecord> mediaRecordsByIds(@Argument final List<UUID> ids, final DataFetchingEnvironment env) {
-        return mediaService.getMediaRecordsByIds(
-                ids,
-                uploadUrlInSelectionSet(env),
-                downloadUrlInSelectionSet(env)
-        );
+    public List<MediaRecord> mediaRecordsByIds(@Argument final List<UUID> ids, final DataFetchingEnvironment env, @ContextValue final LoggedInUser currentUser) {
+        final List<MediaRecord> mediaRecords = mediaService.getMediaRecordsByIds(ids, uploadUrlInSelectionSet(env), downloadUrlInSelectionSet(env));
+
+        return checkAccessForMediaRecords(currentUser, mediaRecords, UserRoleInCourse.STUDENT);
```
Comment (myluki2000) [inline src/main/java/de/unistuttgart/iste/gits/media_service/controller/MediaController.java:52]: should throw an error, not return null

File: src/main/java/de/unistuttgart/iste/gits/media_service/controller/MediaController.java [hunk 4]
```diff
@@ -25,63 +37,55 @@ public class MediaController {
...
     public List<MediaRecord> userMediaRecords(@ContextValue final LoggedInUser currentUser,
                                               final DataFetchingEnvironment env) {
-        return mediaService.getMediaRecordsForUser(
-                currentUser.getId(),
-                uploadUrlInSelectionSet(env),
-                downloadUrlInSelectionSet(env)
-        );
+        final List<MediaRecord> mediaRecords = mediaService.getMediaRecordsForUser(currentUser.getId(), uploadUrlInSelectionSet(env), downloadUrlInSelectionSet(env));
+
+        return checkAccessForMediaRecords(currentUser, mediaRecords, UserRoleInCourse.STUDENT);
```
Comment (myluki2000) [inline src/main/java/de/unistuttgart/iste/gits/media_service/controller/MediaController.java:68]: should throw an error, not return null

File: src/main/java/de/unistuttgart/iste/gits/media_service/controller/MediaController.java [hunk 5]
```diff
@@ -25,63 +37,55 @@ public class MediaController {
...
     public List<List<MediaRecord>> mediaRecordsByContentIds(@Argument final List<UUID> contentIds,
-                                                            final DataFetchingEnvironment env) {
-        return mediaService.getMediaRecordsByContentIds(
-                contentIds,
-                uploadUrlInSelectionSet(env),
-                downloadUrlInSelectionSet(env)
-        );
+                                                            final DataFetchingEnvironment env, @ContextValue final LoggedInUser currentUser) {
+        final List<List<MediaRecord>> mediaRecordsByContentIds = mediaService.getMediaRecordsByContentIds(contentIds, uploadUrlInSelectionSet(env), downloadUrlInSelectionSet(env));
+        return checkAccessForSubLists(currentUser, mediaRecordsByContentIds, UserRoleInCourse.STUDENT);
```
Comment (myluki2000) [inline src/main/java/de/unistuttgart/iste/gits/media_service/controller/MediaController.java:75]: should throw an error, not return null

File: src/main/java/de/unistuttgart/iste/gits/media_service/controller/MediaController.java [hunk 6]
```diff
@@ -25,63 +37,55 @@ public class MediaController {
...
     @QueryMapping
-    public List<List<MediaRecord>> mediaRecordsForCourses(@Argument final List<UUID> courseIds, final DataFetchingEnvironment env) {
-        return mediaService.getMediaRecordsForCourses(
-                courseIds,
-                uploadUrlInSelectionSet(env),
-                downloadUrlInSelectionSet(env)
-        );
+    public List<List<MediaRecord>> mediaRecordsForCourses(@Argument final List<UUID> courseIds, final DataFetchingEnvironment env, @ContextValue final LoggedInUser currentUser) {
+        final List<List<MediaRecord>> mediaRecordsByContentIds = mediaService.getMediaRecordsForCourses(courseIds, uploadUrlInSelectionSet(env), downloadUrlInSelectionSet(env));
+        return checkAccessForSubLists(currentUser, mediaRecordsByContentIds, UserRoleInCourse.STUDENT);
```
Comment (myluki2000) [inline src/main/java/de/unistuttgart/iste/gits/media_service/controller/MediaController.java:81]: should throw an error, not return null

**Review by myluki2000 — CHANGES_REQUESTED** [review]

**Review by myluki2000 — APPROVED** [review]

**Pull request merged**